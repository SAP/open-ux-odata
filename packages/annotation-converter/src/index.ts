export * from './converter';
export * from './utils';
export * from './writeback';
