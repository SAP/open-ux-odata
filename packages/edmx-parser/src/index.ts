export { parse } from './parser';
export { merge } from './merger';
