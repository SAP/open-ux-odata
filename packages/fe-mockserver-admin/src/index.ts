import type { IMockserverPlugin } from '@sap-ux/fe-mockserver-core';
import * as path from 'node:path';

const mockServerAdminConfig: IMockserverPlugin = {
    name: 'mockserver-admin',
    services: [
        {
            alias: 'AdminService',
            urlPath: '/admin',
            metadataPath: path.join(__dirname, 'metadata.xml'),
            mockdataPath: path.join(__dirname, 'mockdata'),
            generateMockData: false
        }
    ]
};

export = mockServerAdminConfig;
