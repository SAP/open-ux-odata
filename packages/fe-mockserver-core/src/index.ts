import type { MockserverConfiguration } from './api';
import Router from 'router';
import type { IRouter } from 'router';
import { createMockMiddleware } from './middleware';
import * as path from 'path';

export interface IFileLoader {
    loadFile(filePath: string): Promise<string>;
    exists(filePath: string): Promise<boolean>;
    loadJS(filePath: string): Promise<any>;
}
export interface IMetadataProcessor {
    loadMetadata(filePath: string): Promise<string>;
}
export * from './api';
export { MockDataContributor } from './mockdata/functionBasedMockData';

export default class FEMockserver {
    isReady: Promise<void>;
    private fileLoader: IFileLoader;
    private metadataProvider: IMetadataProcessor;
    private readonly mainRouter: IRouter;

    constructor(private configuration: MockserverConfiguration) {
        this.mainRouter = new Router();
        this.isReady = this.initialize();
    }

    private async initialize() {
        const FileLoaderClass =
            (this.configuration.fileLoader as any) || (await import('./plugins/fileSystemLoader')).default;
        this.fileLoader = new FileLoaderClass() as IFileLoader;

        const MetadataProviderClass = (
            await this.fileLoader.loadJS(
                this.configuration.metadataProcessor?.name || path.resolve(__dirname, './plugins/metadataProvider')
            )
        ).default;
        this.metadataProvider = new MetadataProviderClass(
            this.fileLoader,
            this.configuration.metadataProcessor?.options
        ) as IMetadataProcessor;
        await createMockMiddleware(this.configuration, this.mainRouter, this.fileLoader, this.metadataProvider);
    }

    getRouter() {
        return this.mainRouter;
    }
}
