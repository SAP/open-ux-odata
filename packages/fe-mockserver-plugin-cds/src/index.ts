import { compileSources, to } from '@sap/cds-compiler';
import { commonCDS } from './common.cds';
import type { IFileLoader, IMetadataProcessor } from '@sap-ux/fe-mockserver-core';
import path from 'path';

export type CDSMetadataProviderOptions = {
    odataVersion?: 'v2' | 'v4';
};

/**
 * Metadata Processor class dedicated to handling cds source files.
 */
export default class CDSMetadataProvider implements IMetadataProcessor {
    /**
     * Constructor for the CDSMetadataProvider.
     *
     * @param fileLoader the file loader injected by the mockserver
     * @param options a set of options for the plugin
     */
    constructor(private fileLoader: IFileLoader, private options?: CDSMetadataProviderOptions) {}

    /**
     * Load the metadata file (cds or xml) and process it accordingly.
     *
     * @param filePath The location of the file
     * @returns a promise with the EDMX content
     */
    async loadMetadata(filePath: string): Promise<string> {
        let cdsContent = await this.fileLoader.loadFile(filePath);
        if (filePath.endsWith('.xml')) {
            return cdsContent;
        }
        const usingReg = new RegExp(/from '([^']+)/g);
        const matches = cdsContent.match(usingReg);
        if (matches) {
            let additionalFiles: string[] = matches.map((match) => /from '([^']+)/.exec(match)![1]);
            additionalFiles = additionalFiles.filter((fileName) => fileName !== '@sap/cds/common');

            for (const additionalFilesKey of additionalFiles) {
                cdsContent += await this.fileLoader.loadFile(path.resolve(path.dirname(filePath), additionalFilesKey));
            }
        }

        const csn = compileSources({ 'string.cds': cdsContent, 'common.cds': commonCDS }, {});
        const edmx = to.edmx.all(csn, {
            odataVersion: this.options?.odataVersion || 'v4',
            odataForeignKeys: true,
            odataContainment: false
        });
        if (Object.keys(edmx).length > 1) {
            throw new Error(`Compilation failed, you can only define one service, found ${Object.keys(edmx).length}`);
        }
        return edmx[Object.keys(edmx)[0]];
    }
}
