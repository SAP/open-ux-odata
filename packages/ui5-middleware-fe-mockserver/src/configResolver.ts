import path from 'path';
import fs from 'fs';
import type {
    ConfigAnnotation,
    ConfigService,
    FileBasedServerConfig,
    FolderBasedServerConfig,
    MockserverConfiguration,
    ServerConfig,
    ServiceConfig
} from '@sap-ux/fe-mockserver-core';

function isFolderBasedConfig(serverConfig: ServerConfig): serverConfig is FolderBasedServerConfig {
    return (serverConfig as FolderBasedServerConfig).mockFolder !== undefined;
}

function isAnnotationConfig(serverConfig: ConfigAnnotation | ConfigService): serverConfig is ConfigAnnotation {
    return (serverConfig as ConfigAnnotation).type?.toLowerCase() === 'annotation';
}

function prepareFolderBasedConfig(
    currentBasePath: string,
    inAnnotations: ConfigAnnotation[],
    inServices: ConfigService[]
) {
    let mockConfig;
    if (fs.existsSync(path.join(currentBasePath, 'config.js'))) {
        mockConfig = require(path.join(currentBasePath, 'config.js'));
    } else {
        mockConfig = JSON.parse(fs.readFileSync(path.join(currentBasePath, 'config.json')).toString('utf-8'));
    }
    mockConfig.forEach((mockConfigEntry: ConfigAnnotation | ConfigService) => {
        if (isAnnotationConfig(mockConfigEntry)) {
            delete mockConfigEntry.type;
            inAnnotations.push(mockConfigEntry);
        } else {
            inServices.push(mockConfigEntry);
        }
    });
}

function prepareFileBasedConfig(inConfig: FileBasedServerConfig) {
    let inServiceFromConfig = inConfig.service ? inConfig.service : inConfig.services;
    if (!Array.isArray(inServiceFromConfig) && inServiceFromConfig !== undefined) {
        inServiceFromConfig = [inServiceFromConfig];
    } else if (inServiceFromConfig === undefined) {
        inServiceFromConfig = [];
    }
    const inServices = inServiceFromConfig as ConfigService[];
    let inAnnotationsFromConfig = inConfig.annotations;
    if (!Array.isArray(inAnnotationsFromConfig) && inAnnotationsFromConfig !== undefined) {
        inAnnotationsFromConfig = [inAnnotationsFromConfig];
    } else if (inAnnotationsFromConfig === undefined) {
        inAnnotationsFromConfig = [];
    }

    const inAnnotations = inAnnotationsFromConfig as ConfigAnnotation[];
    inAnnotations.forEach((annotationConfig: ConfigAnnotation) => annotationConfig.type === 'Annotation');
    return { inServices, inAnnotations };
}

/**
 * Ensure that each service configuration is properly resolved including all file path.
 *
 * @param inServices
 * @param currentBasePath
 * @param inConfig
 * @returns an up to date configuration for one service
 */
function processServicesConfig(
    inServices: ConfigService[],
    currentBasePath: string,
    inConfig: FileBasedServerConfig | FolderBasedServerConfig
) {
    return inServices.map((inService) => {
        const myServiceConfig: ServiceConfig = {
            watch: inService.watch,
            urlPath: inService.urlPath,
            noETag: inService.noETag,
            debug: inService.debug,
            strictKeyMode: inService.strictKeyMode,
            generateMockData: inService.generateMockData,
            contextBasedIsolation: inService.contextBasedIsolation
        } as any;
        const metadataPath = inService.metadataPath || inService.metadataXmlPath || inService.metadataCdsPath;
        if (metadataPath) {
            myServiceConfig.metadataPath = path.resolve(currentBasePath, metadataPath);
        }
        const mockDataPath = inService.mockdataPath || inService.mockdataRootPath;
        if (mockDataPath) {
            myServiceConfig.mockdataPath = path.resolve(currentBasePath, mockDataPath);
        } else {
            // we default to the folder of the metadata
            myServiceConfig.mockdataPath = path.dirname(myServiceConfig.metadataPath);
        }

        if (!inService.urlPath) {
            myServiceConfig.urlPath = inService.urlBasePath + '/' + inService.name;
        }

        if (inConfig.watch && !inService.hasOwnProperty('watch')) {
            myServiceConfig.watch = inConfig.watch;
        }
        if (inConfig.noETag && !inService.hasOwnProperty('noETag')) {
            myServiceConfig.noETag = inConfig.noETag;
        }
        if (inConfig.debug && !inService.hasOwnProperty('debug')) {
            myServiceConfig.debug = inConfig.debug;
        }
        if (inConfig.strictKeyMode && !inService.hasOwnProperty('strictKeyMode')) {
            myServiceConfig.strictKeyMode = inConfig.strictKeyMode;
        }
        if (inConfig.contextBasedIsolation && !inService.hasOwnProperty('contextBasedIsolation')) {
            myServiceConfig.contextBasedIsolation = inConfig.contextBasedIsolation;
        }
        if (inConfig.generateMockData && !inService.hasOwnProperty('generateMockData')) {
            myServiceConfig.generateMockData = inConfig.generateMockData;
        }

        return myServiceConfig;
    });
}

export function resolveConfig(inConfig: ServerConfig, basePath: string): MockserverConfiguration {
    let inServices: ConfigService[] = [];
    let inAnnotations: ConfigAnnotation[] = [];
    let currentBasePath: string = basePath;
    if (isFolderBasedConfig(inConfig)) {
        inConfig.mockFolder = path.resolve(basePath, inConfig.mockFolder);
        currentBasePath = inConfig.mockFolder;
        prepareFolderBasedConfig(currentBasePath, inAnnotations, inServices);
    } else {
        const __ret = prepareFileBasedConfig(inConfig);
        inServices = __ret.inServices;
        inAnnotations = __ret.inAnnotations;
    }
    const annotations = inAnnotations.map((inAnnotation) => {
        inAnnotation.localPath = path.resolve(currentBasePath, inAnnotation.localPath);
        return inAnnotation;
    });
    const services = processServicesConfig(inServices, currentBasePath, inConfig);

    return {
        contextBasedIsolation: !!inConfig.contextBasedIsolation,
        watch: !!inConfig.watch,
        strictKeyMode: !!inConfig.strictKeyMode,
        debug: !!inConfig.debug,
        generateMockData: !!inConfig.generateMockData,
        annotations: annotations,
        services: services,
        fileLoader: inConfig.fileLoader,
        metadataProcessor: inConfig.metadataProcessor
    };
}
