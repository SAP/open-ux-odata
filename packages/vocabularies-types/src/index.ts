export * from './BaseEdm';
export * from './Edm';
