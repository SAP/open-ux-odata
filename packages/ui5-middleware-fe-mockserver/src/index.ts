import * as path from 'path';
import FEMockserver from '@sap-ux/fe-mockserver-core';
import { resolveConfig } from './configResolver';
import type { IRouter } from 'router';
export type { MockDataContributor } from '@sap-ux/fe-mockserver-core';

async function FEMiddleware(middlewareConfig: any): Promise<IRouter> {
    // basepath will be the webapp folder so we have to go up a level to retrieve the config
    const basePath = middlewareConfig.resources
        ? path.resolve(middlewareConfig.resources.rootProject._readers[0]._fsBasePath, '..')
        : '';
    const mockserverInstance = new FEMockserver(resolveConfig(middlewareConfig.options.configuration, basePath));
    await mockserverInstance.isReady;
    return mockserverInstance.getRouter();
}
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export = FEMiddleware;
