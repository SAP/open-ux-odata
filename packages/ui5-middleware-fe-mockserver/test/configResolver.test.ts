import { resolveConfig } from '../src/configResolver';
import path from 'path';

describe('The config resolver', () => {
    it('can resolve the configuration', () => {
        const myBaseResolvedConfig = resolveConfig(
            {
                annotations: {
                    localPath: 'myAnnotation.xml',
                    urlPath: '/my/Annotation.xml'
                },
                service: {
                    urlBasePath: '/my/service',
                    name: 'URL',
                    metadataXmlPath: 'metadata.xml',
                    mockdataRootPath: 'mockData',
                    watch: false,
                    debug: false,
                    contextBasedIsolation: false
                },
                watch: true,
                debug: true,
                strictKeyMode: true,
                contextBasedIsolation: true
            },
            '/'
        );
        expect(myBaseResolvedConfig).toMatchSnapshot();
        let myResolvedConfig = resolveConfig(
            {
                annotations: {
                    localPath: 'myAnnotation.xml',
                    urlPath: '/my/Annotation.xml'
                },
                services: [
                    {
                        urlBasePath: '/my/service',
                        name: 'URL',
                        metadataXmlPath: 'metadata.xml',
                        mockdataRootPath: 'mockData',
                        watch: false,
                        debug: false,
                        contextBasedIsolation: false
                    }
                ],
                watch: true,
                debug: true,
                strictKeyMode: true,
                contextBasedIsolation: true
            },
            '/'
        );
        expect(myResolvedConfig).toEqual(myBaseResolvedConfig);
        myResolvedConfig = resolveConfig(
            {
                annotations: {
                    localPath: 'myAnnotation.xml',
                    urlPath: '/my/Annotation.xml'
                },
                services: [
                    {
                        urlPath: '/my/service/URL',
                        metadataXmlPath: 'metadata.xml',
                        mockdataRootPath: 'mockData',
                        watch: false,
                        debug: false,
                        contextBasedIsolation: false
                    }
                ],
                watch: true,
                debug: true,
                strictKeyMode: true,
                contextBasedIsolation: true
            },
            '/'
        );
        expect(myResolvedConfig).toEqual(myBaseResolvedConfig);
        myResolvedConfig = resolveConfig(
            {
                mockFolder: __dirname,
                watch: true,
                debug: true,
                strictKeyMode: true,
                contextBasedIsolation: true
            },
            '/'
        );
        myResolvedConfig = JSON.parse(JSON.stringify(myResolvedConfig).replace(new RegExp(__dirname, 'g'), ''));
        expect(myResolvedConfig).toEqual(myBaseResolvedConfig);
        myResolvedConfig = resolveConfig(
            {
                mockFolder: path.resolve(__dirname, 'jsConfig'),
                watch: true,
                debug: true,
                strictKeyMode: true,
                contextBasedIsolation: true
            },
            '/'
        );
        myResolvedConfig = JSON.parse(JSON.stringify(myResolvedConfig).replace(new RegExp(__dirname, 'g'), ''));
        expect(myResolvedConfig).toEqual(myBaseResolvedConfig);
    });

    it('can also resolve cds path or empty services', () => {
        const myBaseResolvedConfig = resolveConfig(
            {
                annotations: {
                    localPath: 'myAnnotation.xml',
                    urlPath: '/my/Annotation.xml'
                },
                service: {
                    urlBasePath: '/my/service',
                    name: 'URL',
                    metadataCdsPath: 'metadata.cds',
                    mockdataRootPath: 'mockData'
                }
            },
            '/'
        );
        expect(myBaseResolvedConfig).toMatchSnapshot();

        const myBaseResolvedConfig2 = resolveConfig({}, '/');
        expect(myBaseResolvedConfig2).toMatchSnapshot();
    });

    it('can also apply overrides per service', () => {
        const myBaseResolvedConfig = resolveConfig(
            {
                annotations: {
                    localPath: 'myAnnotation.xml',
                    urlPath: '/my/Annotation.xml'
                },
                services: [
                    {
                        urlBasePath: '/my/service',
                        name: 'URL',
                        metadataCdsPath: 'metadata.cds',
                        mockdataRootPath: 'mockData'
                    },
                    {
                        urlBasePath: '/my/other/service',
                        name: 'URL',
                        metadataCdsPath: 'metadata.cds',
                        mockdataRootPath: 'mockData',
                        watch: false,
                        debug: false,
                        strictKeyMode: false,
                        contextBasedIsolation: false,
                        noETag: false
                    }
                ],
                watch: true,
                debug: true,
                strictKeyMode: true,
                contextBasedIsolation: true,
                noETag: true
            },
            '/'
        );
        expect(myBaseResolvedConfig).toMatchSnapshot();
    });
});
