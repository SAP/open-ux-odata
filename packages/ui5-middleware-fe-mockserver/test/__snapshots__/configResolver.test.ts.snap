// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The config resolver can also apply overrides per service 1`] = `
Object {
  "annotations": Array [
    Object {
      "localPath": "/myAnnotation.xml",
      "urlPath": "/my/Annotation.xml",
    },
  ],
  "contextBasedIsolation": true,
  "debug": true,
  "fileLoader": undefined,
  "metadataProcessor": undefined,
  "services": Array [
    Object {
      "contextBasedIsolation": true,
      "debug": true,
      "metadataPath": "/metadata.cds",
      "mockdataPath": "/mockData",
      "noETag": true,
      "strictKeyMode": true,
      "urlPath": "/my/service/URL",
      "watch": true,
    },
    Object {
      "metadataPath": "/metadata.cds",
      "mockdataPath": "/mockData",
      "urlPath": "/my/other/service/URL",
    },
  ],
  "strictKeyMode": true,
  "watch": true,
}
`;

exports[`The config resolver can also resolve cds path or empty services 1`] = `
Object {
  "annotations": Array [
    Object {
      "localPath": "/myAnnotation.xml",
      "urlPath": "/my/Annotation.xml",
    },
  ],
  "contextBasedIsolation": false,
  "debug": false,
  "fileLoader": undefined,
  "metadataProcessor": undefined,
  "services": Array [
    Object {
      "metadataPath": "/metadata.cds",
      "mockdataPath": "/mockData",
      "urlPath": "/my/service/URL",
    },
  ],
  "strictKeyMode": false,
  "watch": false,
}
`;

exports[`The config resolver can also resolve cds path or empty services 2`] = `
Object {
  "annotations": Array [],
  "contextBasedIsolation": false,
  "debug": false,
  "fileLoader": undefined,
  "metadataProcessor": undefined,
  "services": Array [],
  "strictKeyMode": false,
  "watch": false,
}
`;

exports[`The config resolver can resolve the configuration 1`] = `
Object {
  "annotations": Array [
    Object {
      "localPath": "/myAnnotation.xml",
      "urlPath": "/my/Annotation.xml",
    },
  ],
  "contextBasedIsolation": true,
  "debug": true,
  "fileLoader": undefined,
  "metadataProcessor": undefined,
  "services": Array [
    Object {
      "metadataPath": "/metadata.xml",
      "mockdataPath": "/mockData",
      "strictKeyMode": true,
      "urlPath": "/my/service/URL",
    },
  ],
  "strictKeyMode": true,
  "watch": true,
}
`;
